README


This archive contains the distribution signatures,
version 0.11:

  Subroutine signatures with no source filter

This software is copyright (c) 2008 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.029.



Changes
Revision history for signatures

0.11      2015-01-03 20:35:48Z
  - re-release to fix problematic $VERSION declaration (RT#101095)

0.10      2014-12-22 08:14:23Z
  - fix compile errors on perl 5.21.6 (RT#101076, Father Chrysostomos)

0.09      2014-05-25 16:17:38Z
  - bump required version of B::Hooks::Parser for MSWin32

0.08      2014-05-24 05:07:22Z
  - Patch for perl > 5.19.2 by Peter Martini (RT#95181)

0.07      2012-05-19 10:46:23Z
  - Patch for perl > 5.13.7 by Colin Watson
    See https://rt.cpan.org/Public/Bug/Display.html?id=66997

0.06      2009-07-31 04:34:56Z
  - Require ExtUtils::Depends 0.302.

0.05      2009-04-19 17:21:35Z
  - Depend on B::Hooks::EndOfScope 0.08 so exceptions
    during compile time that trigger a destructor don't
    cause a panic on 5.10 anymore.

0.04      2009-04-19 10:00:18Z
  - Skip a test for syntax errors in string evals that
    sometimes makes perl crash.
  - Add test case for something that used to blow up.
  - Depend on latest B::Hooks::OP::Check.
  - Add repository and bugtracker urls to META.yml.

0.03      2008-12-09 13:35:24Z
  - Suppress a new prototype warning from blead perl.

0.02      2008-11-11 14:26:33Z
  - Port to B::Hooks::OP::PPAddr.

0.01      2008-11-05 08:58:17Z
  - Initial release.


lib/signatures.pm
use strict;
use warnings;
package signatures; # git description: v0.10-1-gac9726f
# ABSTRACT: Subroutine signatures with no source filter
$signatures::VERSION = '0.11';
use XSLoader;
use B::Hooks::Parser 0.12;
use B::Hooks::OP::Check 0.17;
use B::Hooks::OP::PPAddr;
use B::Hooks::EndOfScope 0.08;

XSLoader::load(
    __PACKAGE__,
    exists $signatures::{VERSION}
        ? ${ $signatures::{VERSION} }
        : (),
);

{
    my %pkgs;

    sub import {
        my ($class) = @_;
        my $caller = caller();
        $pkgs{$caller} = $class->setup_for($caller);
        return;
    }

    sub unimport {
        my ($class) = @_;
        my $caller = caller();
        $class->teardown_for(delete $pkgs{$caller});
        return;
    }
}

sub setup_for {
    my ($class, $caller) = @_;
    my $ret = $class->setup($caller);

    $^H{"${class}::enabled"} = 1;

    my $old_warn = $SIG{__WARN__};
    $SIG{__WARN__} = sub {
        if ($_[0] !~ /^(?:(?:Illegal character in prototype)|(?:Prototype after '.')) for /) {
            $old_warn ? $old_warn->(@_) : warn @_;
        }
    };

    my $unregister;
    {
        my $called = 0;
        $unregister = sub {
            return if $called++;
            $class->teardown_for([$ret, $unregister]);
            $SIG{__WARN__} = $old_warn;
        };
    }

    &on_scope_end($unregister);

    return [$ret, $unregister];
}

sub teardown_for {
    my ($class, $data) = @_;
    delete $^H{"${class}::enabled"};
    $class->teardown($data->[0]);
    $data->[1]->();
    return;
}

sub callback {
    my ($class, $offset, $proto) = @_;
    my $inject = $class->proto_unwrap($proto);
    $class->inject($offset, $inject);
    return;
}

sub proto_unwrap {
    my ($class, $proto) = @_;
    return '' unless length $proto;
    return "my ($proto) = \@_;";
}

sub inject {
    my ($class, $offset, $inject) = @_;
    my $linestr = B::Hooks::Parser::get_linestr();
    substr($linestr, $offset + 1, 0) = $inject;
    B::Hooks::Parser::set_linestr($linestr);
    return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

signatures - Subroutine signatures with no source filter

=head1 VERSION

version 0.11

=head1 SYNOPSIS

    use signatures;

    sub foo ($bar, $baz) {
        return $bar + $baz;
    }

=head1 DESCRIPTION

With this module, we can specify subroutine signatures and have variables
automatically defined within the subroutine.

For example, you can write

    sub square ($num) {
        return $num * $num;
    }

and it will be automatically turned into the following at compile time:

    sub square {
        my ($num) = @_;
        return $num * $num;
    }

Note that, although the syntax is very similar, the signatures provided by this
module are not to be confused with the prototypes described in L<perlsub>. All
this module does is extracting items of @_ and assigning them to the variables
in the parameter list. No argument validation is done at runtime.

The signature definition needs to be on a single line only.

If you want to combine sub signatures with regular prototypes a C<proto>
attribute exists:

    sub foo ($bar, $baz) : proto($$) { ... }

=head1 METHODS

If you want subroutine signatures doing something that this module doesn't
provide, like argument validation, typechecking and similar, you can subclass
it and override the following methods.

=head2 proto_unwrap ($prototype)

Turns the extracted C<$prototype> into code.

The default implementation returns C<< my (${prototype}) = @_; >> or an empty
string, if no prototype is given.

=head2 inject ($offset, $code)

Inserts a C<$code> string into the line perl currently parses at the given
C<$offset>. This is only called by the C<callback> method.

=head2 callback ($offset, $prototype)

This gets called as soon as a sub definition with a prototype is
encountered. Arguments are the C<$offset> within the current line perl
is parsing and extracted C<$prototype>.

The default implementation calls C<proto_unwrap> with the prototype and passes
the returned value and the offset to C<inject>.

=for Pod::Coverage setup setup_for teardown teardown_for

=head1 BUGS

=over 4

=item prototypes aren't checked for validity yet

You won't get a warning for invalid prototypes using the C<proto> attribute,
like you normally would with warnings enabled.

=item you shouldn't alter $SIG{__WARN__} at compile time

After this module is loaded you shouldn't make any changes to C<$SIG{__WARN__}>
during compile time. Changing it before the module is loaded or at runtime is
fine.

=back

=head1 SEE ALSO

L<Method::Signatures>

L<MooseX::Method::Signatures>

L<Sub::Signatures>

L<Attribute::Signature>

L<Perl6::Subs>

L<Perl6::Parameters>

=head1 THANKS

=for stopwords Moritz Lenz Steffen Schwigon

Moritz Lenz and Steffen Schwigon for documentation review and
improvement.

=head1 AUTHOR

Florian Ragwitz <rafl@debian.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 CONTRIBUTORS

=for stopwords Karen Etheridge Peter Martini Alexandr Ciornii Father Chrysostomos Alex Kapranoff Steffen Schwigon

=over 4

=item *

Karen Etheridge <ether@cpan.org>

=item *

Peter Martini <PeterCMartini@GMail.com>

=item *

Alexandr Ciornii <alexchorny@gmail.com>

=item *

Father Chrysostomos <sprout@cpan.org>

=item *

Alex Kapranoff <alex@kapranoff.ru>

=item *

Steffen Schwigon <ss5@renormalist.net>

=back

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Subroutine signatures with no source filter",
  "AUTHOR" => "Florian Ragwitz <rafl\@debian.org>",
  "BUILD_REQUIRES" => {
    "B::Hooks::OP::Check" => "0.17",
    "B::Hooks::OP::PPAddr" => "0.03",
    "B::Hooks::Parser" => "0.12",
    "ExtUtils::Depends" => "0.302"
  },
  "CONFIGURE_REQUIRES" => {
    "B::Hooks::OP::Check" => "0.17",
    "B::Hooks::OP::PPAddr" => "0.03",
    "B::Hooks::Parser" => "0.12",
    "ExtUtils::Depends" => "0.302",
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "signatures",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "signatures",
  "PREREQ_PM" => {
    "B::Hooks::EndOfScope" => "0.08",
    "B::Hooks::OP::Check" => "0.17",
    "B::Hooks::OP::PPAddr" => "0.03",
    "B::Hooks::Parser" => "0.12",
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "base" => 0,
    "vars" => 0
  },
  "VERSION" => "0.11",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

use ExtUtils::Depends 0.302;
my $pkg = ExtUtils::Depends->new(
    'signatures',
    'B::Hooks::OP::Check',
    'B::Hooks::OP::PPAddr',
    'B::Hooks::Parser',
);
$pkg->add_xs('signatures.xs');
$pkg->add_pm('lib/signatures.pm' => '$(INST_LIB)/signatures.pm');
%WriteMakefileArgs = ( %WriteMakefileArgs, $pkg->get_makefile_vars );




my %FallbackPrereqs = (
  "B::Hooks::EndOfScope" => "0.08",
  "B::Hooks::OP::Check" => "0.17",
  "B::Hooks::OP::PPAddr" => "0.03",
  "B::Hooks::Parser" => "0.12",
  "ExtUtils::Depends" => "0.302",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "XSLoader" => 0,
  "base" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


